session/src/main/java/org/apache/iotdb/session/template/Template.java

```diff
@@ -0,0 +1,241 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.session.template;
+
+import org.apache.iotdb.rpc.StatementExecutionException;
+import org.apache.iotdb.tsfile.common.constant.TsFileConstant;
+import org.apache.iotdb.tsfile.file.metadata.enums.CompressionType;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSEncoding;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+
+public class Template {
+  private String name;
+  private Node moutedNode;
+  private boolean shareTime;
+
+  private List<String> measurementsPath;
+  private List<Boolean> alignedList;
+  private List<TSDataType> dataTypes;
+  private List<TSEncoding> encodings;
+  private List<CompressionType> compressors;
+
+  // sync with metadata.Template
+  public enum TemplateQueryType {
+    NULL,
+    COUNT_MEA,
+    IS_MEA,
+    IS_SERIES,
+    SHOW_MEA
+  }
+
+  public Template(String name, boolean isShareTime) {
+    this.name = name;
+    moutedNode = new InternalNode(name, isShareTime);
+    this.shareTime = isShareTime;
+
+    this.measurementsPath = new ArrayList<>();
+    this.alignedList = new ArrayList<>();
+    this.dataTypes = new ArrayList<>();
+    this.encodings = new ArrayList<>();
+    this.compressors = new ArrayList<>();
+  }
+
+  public Template(String name) {
+    this(name, false);
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public boolean isShareTime() {
+    return shareTime;
+  }
+
+  public void setShareTime(boolean shareTime) {
+    this.shareTime = shareTime;
+  }
+
+  public List<Boolean> getAlignedList() {
+    return this.alignedList;
+  }
+
+  public List<String> getMeasurementsPath() {
+    return this.measurementsPath;
+  }
+
+  public List<TSEncoding> getEncodings() {
+    return encodings;
+  }
+
+  public List<CompressionType> getCompressors() {
+    return compressors;
+  }
+
+  // region Interface to manipulate Template
+
+  public void addToTemplate(Node child) throws StatementExecutionException {
+    if (moutedNode.getChildren().containsKey(child.getName()))
+      throw new StatementExecutionException("Duplicated child of node in template.");
+    moutedNode.getChildren().put(child.getName(), child);
```

Comment:
how could you construct a tree using a non-recursion method? This should be a bug